code/modules/mining/ambush_structures.dm

```diff
@@ -0,0 +1,139 @@
+
+///now we actually do the heavy lifting.
+/obj/structure/ambush_burrow/proc/spawn_mobs()
+
+	shake_animation(intensity = 14)//shake again for good measure
+	playsound(src, pick(crumble_sound), 40)
+	icon_state = "maint_hole"
+	//get potential directions to place a mob
+	var/list/possible_directions = GLOB.cardinal.Copy()
+	var/mobs_spawned = 0
+	var/probability = our_ambush.special_probability
+
+	while(mobs_spawned < our_ambush.mob_spawn && possible_directions.len)
+		var/turf/possible_T = get_step(loc, pick_n_take(possible_directions))
+		new /obj/effect/decal/cleanable/rubble(possible_T)
+		var/mobtype
+		if(prob(probability))//if prob allows, pick a special mob
+			mobtype = pick(our_ambush.special_types)
+			//proba = max(0, probability - 5)  // Decreasing probability to avoid mass spam of special mobs
+		else
+			mobtype = pick(our_ambush.normal_types)  // Pick a normal mob
+		if(!controller.check_density_no_mobs(possible_T))
+			new mobtype(possible_T)  // Spawn mob at free location
+		else
+			new mobtype(loc)
+		mobs_spawned++
+	if(our_ambush.ambush_type == AMBUSH_SKIRMISH)
+		addtimer(CALLBACK(src, PROC_REF(crumble)), 2 SECONDS)
+
+///visibly indicates to players that this burrow is out of commission. Also preps for deletion
+/obj/structure/ambush_burrow/proc/crumble()
+	icon_state = "maint_hole_collapsed"
+	desc = "It's filled with lose debris. As you watch, it begins to crumble away..."
+	addtimer(CALLBACK(src, PROC_REF(stop)), 5 SECOND)
+
+/obj/structure/ambush_burrow/Destroy()
+	visible_message(SPAN_DANGER("\The [src] crumbles away!"))
+	new /obj/effect/decal/cleanable/rubble(src.loc)
+	if(controller)
+		controller.burrows -= src
+		controller = null
+	..()
+
+/obj/structure/ambush_burrow/attack_generic(mob/user, damage)
+	user.do_attack_animation(src)
+	visible_message(SPAN_DANGER("\The [user] smashes \the [src]!"))
+	take_damage(damage)
+	user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN * 1.5)
+
+/obj/structure/ambush_burrow/attackby(obj/item/I, mob/user)
+	if (user.a_intent == I_HURT && user.Adjacent(src))
+		if(!(I.flags & NOBLUDGEON))
+			user.do_attack_animation(src)
+			var/damage = I.force * I.structure_damage_factor
+			var/volume =  min(damage * 3.5, 15)
+			if (I.hitsound)
+				playsound(src, I.hitsound, volume, 1, -1)
+			visible_message(SPAN_DANGER("[src] has been hit by [user] with [I]."))
+			take_damage(damage)
+			user.setClickCooldown(DEFAULT_ATTACK_COOLDOWN * 1.5)
+	return TRUE
+
+/obj/structure/ambush_burrow/bullet_act(obj/item/projectile/Proj)
+	..()
+        // Bullet not really efficient against a pile of debris
+	take_damage(Proj.get_structure_damage() * 0.25)
+
+/obj/structure/ambush_burrow/take_damage(damage)
+	. = health - damage < 0 ? damage - (damage - health) : damage
+	. *= explosion_coverage
+	health = min(max(health - damage, 0), maxHealth)
+	if(health == 0)
+		qdel(src)
+	return
+
+/obj/structure/ambush_burrow/proc/stop()
+	qdel(src)  // Delete burrow
```

Comment:
do you have future plans for this to do more than qdeling? if not, please delete it and just call qdel instead of abstracting it